component/loki/process/stages/pipeline.go

```diff
@@ -0,0 +1,198 @@
+package stages
+
+// This package is ported over from grafana/loki/clients/pkg/logentry/stages.
+// We aim to port the stages in steps, to avoid introducing huge amounts of
+// new code without being able to slowly review, examine and test them.
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component/common/loki"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/time/rate"
+)
+
+// StageConfig defines a single stage in a processing pipeline.
+type StageConfig struct {
+	JSONConfig   *JSONConfig         `river:"json,block,optional"`
+	LabelsConfig *map[string]*string `river:"labels,attr,optional"`
```

Comment:
It's a little weird to me that `json` is a block but `labels` is an attribute. I think this will make the documentation a little confusing to follow, since users will have to be aware that not all stages are subblocks. 

I'd suggest we do something like 

```
loki.process "example" {
  stage {
    labels {
      values = { ... } 
    }
  }
} 
```

just for consistency with the other stages, even though it's kind of ugly and verbose.

Reply:
I also had the same dillemma here, not particularly fond of either solution, but as you said, let's keep the consistency.